### PR TITLE
2 task modification

### PR DESCRIPTION
**Pull request by POPPININ**
Added the files required to facilitate task modification. Test coverage for Controller, Interactor and Presenter. Data Access required data persistence and hence is yet to be implemented.

### CODE REVIEW

**Review by missionStrawberry — CHANGES_REQUESTED** [review]
There are some issues with following the file naming and documentation conventions we set as a team. One major issue is that your use case is missing the view/screen. In your use case that would be the form the user sees when they are modifying a task. There are other issues which I have noted throughout about SOLID principle violation and clean architecture. Another issue of note is that I found a typo in the interaction that results in a logic error. Otherwise good job!

File: src/main/java/entities/TaskFactory.java [hunk 1]
```diff
@@ -17,4 +17,24 @@ public class TaskFactory {
...
+    /**
+     * Modify a given Task by changing its title.
+     *
+     * Precondition: newTitle is a non-empty string, and not equal to any other task
+     * scheduled for the same day.
+     *
+     * @param newTitle title of the task
+     * @param task the underlying Task object
+     */
+    public void modifyTask(String newTitle, Task task) { task.setTitle(newTitle); }
```
[the remark this answers is not in the document]
Comment (DanielLivshits) [reply src/main/java/entities/TaskFactory.java:31]: The factory only appears in the creation use cases. After that, the information from the task is distributed and stored in the view model and data access. The factory only serves to create tasks (as the name "factory" might suggest).

**Review by Fard-Faru — CHANGES_REQUESTED** [review]

File: src/main/java/entities/TaskFactory.java [hunk 2]
```diff
@@ -17,4 +17,24 @@ public class TaskFactory {
...
+
+    /**
+     * Modify a given Task by changing its title.
+     *
+     * Precondition: newTitle is a non-empty string, and not equal to any other task
+     * scheduled for the same day.
+     *
+     * @param newTitle title of the task
+     * @param task the underlying Task object
+     */
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/entities/TaskFactory.java:30]: Added file headers.

File: src/main/java/task_modification_use_case/TaskModificationDataAccess.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package task_modification_use_case;
+
+public class TaskModificationDataAccess implements TaskModificationDsGateway {
+
+    public TaskModificationDataAccess() {}
+
+    @Override
+
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationDataAccess.java:8]: Removed.

File: src/main/java/task_modification_use_case/TaskModificationController.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+public class TaskModificationController {
+    private TaskModificationInputBoundary taskModificationInputBoundary;
+
+    public TaskModificationController(TaskModificationInteractor taskModificationInputBoundary) {
+        this.taskModificationInputBoundary = taskModificationInputBoundary;
+    }
+
+    TaskModificationResponseModel modifyTask(int dayId, String newTitle, String title) {
+        TaskModificationRequestModel requestModel = new TaskModificationRequestModel(
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationController.java]: Changed RequestModel and ResponseModel to InputData and OutputData where applicable.

File: src/main/java/entities/TaskFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/entities/TaskFactory.java]: Removed any dependency on Factory. The business logic now relies on data access (and view model, which is yet to be implemented).

File: src/main/java/task_modification_use_case/TaskModificationController.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package task_modification_use_case;
+
+public class TaskModificationController {
+    private TaskModificationInputBoundary taskModificationInputBoundary;
+
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationController.java:12]: Fixed the logic error in the interactor, which was searching for a task with the old key instead of the key that we wanted to change the title of a given task too. I have shortened the instance variable names where appropriate. View Model will be implemented soon. I am unsure of what other SOLID violations you are talking about, and would appreciate some clarification.

File: src/main/java/task_modification_use_case/TaskModificationController.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    public TaskModificationController(TaskModificationInteractor taskModificationInputBoundary) {
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationController.java]: Fixed this issue across all classes where interface instances were explicitly being initialized with instances of classes which implement those interfaces, as indicated by constructor signature.

File: src/main/java/task_modification_use_case/TaskModificationController.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package task_modification_use_case;
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationController.java:8]: Added file headers to all files.

File: src/main/java/task_modification_use_case/TaskModificationResponseModel.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    String getTitle() { return this.title; }
+
+    void setTitle(String title) { this.title = title; }
+
+    int getDayId() { return this.dayId; }
+
+    public void setDayId(int dayId) { this.dayId = dayId; }
+}
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationResponseModel.java]: Removed setters and getters.

File: src/test/java/task_modification_use_case/TaskModificationControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+                outputBoundary, dsGateway
+        );
+        TaskModificationController controller = new TaskModificationController(inputBoundary);
+        TaskModificationResponseModel responseModel = controller.modifyTask(
+                dayId, newTitle, title
+        );
+        assertEquals(0, responseModel.getDayId());
+        assertEquals("New Sample Task", responseModel.getTitle());
+    }
+}
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationControllerTest.java:26]: Data Access has now been implemented and the relevant failing condition was both identified and corrected.

File: src/test/java/task_modification_use_case/TaskModificationPresenterTest.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package task_modification_use_case;
+import org.junit.jupiter.api.Test;
+
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class TaskModificationPresenterTest {
+    //task and event instances (and maps) to be used across tests.
+    private final static TaskModificationResponseModel task = new TaskModificationResponseModel(
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationPresenterTest.java]: Changed the name of the variable "task" to "response" instead.

File: src/test/java/task_modification_use_case/TaskModificationPresenterTest.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        TaskModificationOutputBoundary outputBoundary = new TaskModificationPresenter();
+        TaskModificationResponseModel responseModel = outputBoundary.prepareSuccessView(
+                task
+        );
+        assertEquals(task.getDayId(), responseModel.getDayId());
+        assertEquals(task.getTitle(), responseModel.getTitle());
+    }
+
+    @Test
+    public void testPrepareFailView(){
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationPresenterTest.java:25]: Would appreciate clarification on what you believe the PresenterTest file should be tesing?

File: src/main/java/entities/TaskFactory.java [hunk 12]
```diff
@@ -17,4 +17,24 @@ public class TaskFactory {
...
+    public void modifyTask(String newTitle, Task task) { task.setTitle(newTitle); }
+
+    /**
+     * Mark a given Task as completed.
+     * @param task the underlying Task object
+     */
+    public void markTaskAsCompleted(Task task) { task.setCompleted(true); }
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/entities/TaskFactory.java]: Yes it is. I will link him to that issue.

File: src/main/java/task_modification_use_case/TaskModificationDataAccess.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+package task_modification_use_case;
+
+import data_access.DataAccessDay;
+import data_access.DataAccessTask;
+import data_access.WeekDataAccess;
+import java.util.HashMap;
+
+public class TaskModificationDataAccess extends WeekDataAccess implements TaskModificationDsGateway {
+
+    public TaskModificationDataAccess() {}
```
Comment (Fard-Faru) [inline src/main/java/task_modification_use_case/TaskModificationDataAccess.java:17]: Why is this included if it is not doing anything?

File: src/test/java/task_modification_use_case/TaskModificationControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+                outputBoundary, dsGateway
+        );
+        TaskModificationController controller = new TaskModificationController(inputBoundary);
+        TaskModificationOutputData outputData = controller.modifyTask(
+                dayId, newTitle, title
+        );
+        assertEquals(0, outputData.getDayId());
+        assertEquals("New Sample Task", outputData.getTitle());
+    }
+}
```
Comment (Fard-Faru) [inline src/test/java/task_modification_use_case/TaskModificationControllerTest.java:26]: Please add a new line at the end of the file to keep a nice code structure and maintain CLEAN consistency across everyone's files.

File: src/test/java/task_modification_use_case/TaskModificationDataAccessTest.java [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package task_modification_use_case;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class TaskModificationDataAccessTest {
+// TODO: Add test cases for when methods in DataAccess have non-trivial functionality
+}
```
Comment (Fard-Faru) [inline src/test/java/task_modification_use_case/TaskModificationDataAccessTest.java:7]: Please add a new line at the end of the file to keep a nice code structure and maintain CLEAN consistency across everyone's files.

File: src/test/java/task_modification_use_case/TaskModificationInteractorTest.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        TaskModificationOutputData outputData = inputBoundary.modifyTask(
+                inputData
+        );
+
+        assertEquals("New Sample Task", outputData.getTitle());
+        assertEquals(0, outputData.getDayId());
+        assertEquals("New Sample Task", task.getTitle());
+    }
+
+}
```
Comment (Fard-Faru) [inline src/test/java/task_modification_use_case/TaskModificationInteractorTest.java:42]: Please add a new line at the end of the file to keep a nice code structure and maintain CLEAN consistency across everyone's files.

**Review by Fard-Faru — CHANGES_REQUESTED** [review]
Hi, Raghav please implement the changes I suggested or feel free to let me know if you disagree with something.

File: src/main/java/task_modification_use_case/TaskModificationDataAccess.java (hunk 13 above)
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationDataAccess.java:17]: Removed the redundant constructor.

File: src/test/java/task_modification_use_case/TaskModificationControllerTest.java (hunk 14 above)
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationControllerTest.java]: Done

File: src/test/java/task_modification_use_case/TaskModificationDataAccessTest.java (hunk 15 above)
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationDataAccessTest.java]: Done.

File: src/test/java/task_modification_use_case/TaskModificationInteractorTest.java (hunk 16 above)
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationInteractorTest.java]: Done

File: src/main/java/task_modification_use_case/TaskModificationController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (missionStrawberry) [reply src/main/java/task_modification_use_case/TaskModificationController.java:12]: The SOLID violation I was talking about was in TaskFactory.

File: src/test/java/task_modification_use_case/TaskModificationPresenterTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (missionStrawberry) [reply src/test/java/task_modification_use_case/TaskModificationPresenterTest.java:25]: I believe what I meant was that because the presenter gets passed in the OutputData, it needs to check that said OutputData was modified in whatever way was necessary rather than check if it created an OutputData

File: src/test/java/task_modification_use_case/TaskModificationPresenterTest.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Test
+    public void testPrepareSuccessView(){
+        TaskModificationOutputBoundary outputBoundary = new TaskModificationPresenter();
+        TaskModificationOutputData response = outputBoundary.prepareSuccessView(
+                outputData
+        );
+        assertEquals(outputData.getDayId(), response.getDayId());
+        assertEquals(outputData.getTitle(), response.getTitle());
+    }
```
Comment (missionStrawberry) [inline src/test/java/task_modification_use_case/TaskModificationPresenterTest.java:22]: check the assert with 0, response.getDayId() and for the second "Sample Task", response.getTitle()

**Review by missionStrawberry — CHANGES_REQUESTED** [review]
Looks much better Raghav! Noted two minor things and then I believe you should be good to go.

File: src/main/java/task_modification_use_case/TaskModificationController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationController.java:12]: The methods in TaskFactory which violated SOLID principles have been removed.

File: src/test/java/task_modification_use_case/TaskModificationPresenterTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationPresenterTest.java:25]: I have addressed this in the latest commit.

File: src/test/java/task_modification_use_case/TaskModificationPresenterTest.java (hunk 17 above)
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationPresenterTest.java]: Done. Explicitly comparing actual values of instance attributes against expected.

File: src/main/java/task_modification_use_case/TaskModificationPresenter.java [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public class TaskModificationPresenter implements TaskModificationOutputBoundary {
+    @Override
+    public TaskModificationOutputData prepareSuccessView(TaskModificationOutputData task) {
+        TaskModificationOutputData outputData = new TaskModificationOutputData(
+                task.getTitle(), task.getDayId()
+        );
```
[the remark this answers is not in the document]
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationPresenter.java]: Removed the redundancy. The method now returns task.

**Review by missionStrawberry — DISMISSED** [review]
Looking great!

File: src/main/java/task_modification_use_case/TaskModificationController.java [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+    private TaskModificationInputBoundary inputBoundary;
+
+    public TaskModificationController(TaskModificationInputBoundary inputBoundary) {
+        this.inputBoundary = inputBoundary;
+    }
+
+    TaskModificationOutputData modifyTask(int dayId, String newTitle, String title) {
+        TaskModificationInputData inputData = new TaskModificationInputData(
+                dayId, newTitle, title
+        );
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationController.java:20]: Following our established naming conventions, this class should be called ModifyTaskController instead of TaskModificationController.

File: src/main/java/task_modification_use_case/TaskModificationDataAccess.java [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+import data_access.DataAccessDay;
+import data_access.DataAccessTask;
+import data_access.WeekDataAccess;
+import java.util.HashMap;
+
+public class TaskModificationDataAccess extends WeekDataAccess implements TaskModificationDsGateway {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationDataAccess.java:15]: Likewise, should be called ModifyTaskDataAccess to follow naming conventions.

File: src/main/java/task_modification_use_case/TaskModificationDsGateway.java [hunk 21]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Task Modification DsGateway Interface.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public interface TaskModificationDsGateway {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationDsGateway.java:10]: Should be called ModifyTaskDsGateway.

File: src/main/java/task_modification_use_case/TaskModificationDsInputData.java [hunk 22]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Task Modification DsInputData Class.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public class TaskModificationDsInputData {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationDsInputData.java:10]: Should be ModifyTaskDsInputData

File: src/main/java/task_modification_use_case/TaskModificationInputBoundary.java [hunk 23]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Task Modification Input Boundary Interface.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public interface TaskModificationInputBoundary {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationInputBoundary.java:10]: Should be ModifyTaskInputBoundary

File: src/main/java/task_modification_use_case/TaskModificationInputData.java [hunk 24]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Task Modification Input Data Class.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public class TaskModificationInputData {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationInputData.java:10]: ModifyTaskInputData

File: src/main/java/task_modification_use_case/TaskModificationInteractor.java [hunk 25]
```diff
@@ -0,0 +1,40 @@
+/**
+ * Task Modification Interactor Class.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public class TaskModificationInteractor implements TaskModificationInputBoundary {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationInteractor.java:10]: ModifyTaskInteractor

File: src/main/java/task_modification_use_case/TaskModificationOutputBoundary.java [hunk 26]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Task Modification Output Boundary Interface.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public interface TaskModificationOutputBoundary {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationOutputBoundary.java:10]: ModifyTaskOutputBoundary

File: src/main/java/task_modification_use_case/TaskModificationOutputData.java [hunk 27]
```diff
@@ -0,0 +1,22 @@
+/**
+ * Task Modification Output Data Class.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public class TaskModificationOutputData {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationOutputData.java:10]: ModifyTaskOutputData

File: src/main/java/task_modification_use_case/TaskModificationPresenter.java [hunk 28]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Task Modification Presenter Class.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
+
+public class TaskModificationPresenter implements TaskModificationOutputBoundary {
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationPresenter.java:10]: ModifyTaskPresenter

File: src/test/java/task_modification_use_case/TaskModificationControllerTest.java [hunk 29]
```diff
@@ -0,0 +1,26 @@
+package task_modification_use_case;
+
+import org.junit.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class TaskModificationControllerTest {
```
Comment (DanielLivshits) [inline src/test/java/task_modification_use_case/TaskModificationControllerTest.java:7]: ModifyTaskControllerTest

File: src/test/java/task_modification_use_case/TaskModificationInteractorTest.java [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+import data_access.DataAccessEvent;
+import data_access.DataAccessTask;
+import data_access.WeekDataAccess;
+
+import java.util.HashMap;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class TaskModificationInteractorTest extends WeekDataAccess {
```
Comment (DanielLivshits) [inline src/test/java/task_modification_use_case/TaskModificationInteractorTest.java:13]: ModifyTaskInteractorTest

File: src/test/java/task_modification_use_case/TaskModificationPresenterTest.java [hunk 31]
```diff
@@ -0,0 +1,35 @@
+package task_modification_use_case;
+import org.junit.jupiter.api.Test;
+
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class TaskModificationPresenterTest {
```
Comment (DanielLivshits) [inline src/test/java/task_modification_use_case/TaskModificationPresenterTest.java:7]: ModifyTaskPresenterTest

File: src/main/java/task_modification_use_case/TaskModificationController.java [hunk 32]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Task Modification Controller Class.
+ * Author: Raghav Arora
+ * Modified By: Raghav Arora
+ * Created: Nov 12, 2022
+ * Last Modified: Nov 19, 2022
+ */
+package task_modification_use_case;
```
Comment (DanielLivshits) [inline src/main/java/task_modification_use_case/TaskModificationController.java:8]: also, the package should be renamed to modify_task_use_case.

**Review by DanielLivshits — CHANGES_REQUESTED** [review]
In general, classes named TaskModificationClass should be renamed to ModifyTaskClass to follow group naming conventions. The package should also be renamed to modify_task_use_case. 

If you didn't already know, IntelliJ has an option to change all references to the name of the class within the project (you can access this through Right Clicking the class file and going to Refactor -> Rename and selecting the appropriate options. I've attached a file below with the screen that should display for this. 

image

File: src/main/java/task_modification_use_case/TaskModificationController.java (hunk 32 above)
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationController.java]: Fixed.

File: src/test/java/task_modification_use_case/TaskModificationPresenterTest.java (hunk 31 above)
Comment (POPPININ) [reply src/test/java/task_modification_use_case/TaskModificationPresenterTest.java]: Fixed.

File: src/main/java/task_modification_use_case/TaskModificationController.java (hunk 19 above)
Comment (POPPININ) [reply src/main/java/task_modification_use_case/TaskModificationController.java]: Refactored package and class names to comply with our convention.

**Review by missionStrawberry — APPROVED** [review]

**Review by DanielLivshits — APPROVED** [review]

**Review by Fard-Faru — APPROVED** [review]
Good Work Team!
